presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -256,37 +291,42 @@ private void processNewSources()
 
     public ListenableFuture<?> process()
     {
+        checkState(initialized.get(), "Driver is not initialized");
         checkLockNotHeld("Can not process while holding the driver lock");
 
+        // if the driver is blocked we don't need to continue
+        synchronized (driverBlockedFutureLock) {
+            if (!driverBlockedFuture.isDone()) {
+                return driverBlockedFuture;
+            }
+        }
+
         Optional<ListenableFuture<?>> result = tryWithLock(100, TimeUnit.MILLISECONDS, () -> {
-            ListenableFuture<?> driverBlockedFuture = processInternal();
-            return wakeUpOnRevokeRequest(driverBlockedFuture);
+            ListenableFuture<?> sourceBlockedFuture = processInternal();
+            synchronized (driverBlockedFutureLock) {
+                updateDriverBlockedFuture(sourceBlockedFuture);
+                return driverBlockedFuture;
+            }
         });
         return result.orElse(NOT_BLOCKED);
     }
 
-    private ListenableFuture<?> wakeUpOnRevokeRequest(ListenableFuture<?> driverBlockedFuture)
+    @GuardedBy("driverBlockedFutureLock")
+    private void updateDriverBlockedFuture(ListenableFuture<?> sourceBlockedFuture)
     {
-        if (driverBlockedFuture.isDone()) {
-            // driver is not blocked; just return completed future
-            return driverBlockedFuture;
-        }
-
-        ImmutableList<SettableFuture<?>> revokingRequestedFutures = operators.stream()
-                .map(Operator::getOperatorContext)
-                .map(OperatorContext::getMemoryRevokingRequestedFuture)
-                .collect(toImmutableList());
-        Optional<SettableFuture<?>> doneRevokingRequestedFuture = revokingRequestedFutures.stream().filter(Future::isDone).findAny();
-        if (doneRevokingRequestedFuture.isPresent()) {
-            // revoking already requested for one of operators; return completed future
-            return doneRevokingRequestedFuture.get();
-        }
-
-        // unblock as soon as driver is unblocked or we get revoking request for any of operators
-        ImmutableList.Builder<ListenableFuture<?>> futures = ImmutableList.builder();
-        futures.add(driverBlockedFuture);
-        futures.addAll(revokingRequestedFutures);
-        return firstFinishedFuture(futures.build());
+        checkState(initialized.get(), "Driver is not initialized");
+        checkState(Thread.holdsLock(driverBlockedFutureLock), "driverBlockedFutureLock must be held to call updateDriverBlockedFuture");
+        // driverBlockedFuture will be completed as soon as the sourceBlockedFuture is completed
+        // or any of the operators gets a memory revocation request
+        driverBlockedFuture = SettableFuture.create();
+        sourceBlockedFuture.addListener(this::completeDriverBlockedFuture, newDirectExecutorService());
```

Comment:
No need to use `completeDriverBlockedFuture` here, is there any?
I think the intention is "when `sourceBlockedFuture` completes,  this particular new `driverBlockedFuture` should complete", so you could just keep the new `driverBlockedFuture` on a variable and complete that, without taking any locks.

Reply:
Also, if you do that, you can replace driverBlockedFutureLock with an AtomicReference and remove all synchronization.